### PR TITLE
Save and load history (issue 96)

### PR DESCRIPTION
**Pull request by alixdamman**
Not finished but I have something working for the "save" part. 
I start the PR to start the discussion about implementation.

### CODE REVIEW

**Comment by alixdamman** [comment]
@gdementen Do you know how to simulation/force return in interactive shell? 
When I try to load a file via the Menu, I have to press Enter twice before to get something. 
Functions Save/Load history "works" but not as I expected.

**Comment by gdementen** [comment]
No, I do not know. I had already tried to do this a while ago (see the comment at https://github.com/larray-project/larray-editor/blob/master/larray_editor/editor.py#L460). I guess we have to either ask ipython devs or dig into the code ourselves to see if it is at all possible. I am not hopeful it is possible to do using a clean/documented/public API (last time, I already searched google & the API reference for quite a while) but it might be possible to abuse the system to achieve that anyway.

File: larray_editor/editor.py [hunk 1]
```diff
@@ -554,6 +586,177 @@ def new(self):
...
+                browse_edit.setText(filepath)
+            browse_filedialog = QFileDialog(self, filter="Python Script (*.py)")
+            browse_filedialog.setFileMode(QFileDialog.ExistingFile)
+            browse_button.clicked.connect(browse_filedialog.open)
+            browse_filedialog.fileSelected.connect(browse_edit.setText)
+            layout.addWidget(browse_label, 0, 0)
+            layout.addWidget(browse_edit, 0, 1)
+            layout.addWidget(browse_button, 0, 2)
+            # lines
+            lines_label = QLabel("Lines")
```
Comment (gdementen) [inline larray_editor/editor.py]: I would rather have the lines_edit and symbols_edit "guarded" by a radio button with "all lines", "lines [line_edit]", and "variables [symbols_edit]" in the same vein than the pages selection in the print dialog of firefox or acrobat reader.
We would loose the possibility to load specific symbols within specific lines, but I guess that would be very rarely used.

File: larray_editor/editor.py [hunk 2]
```diff
@@ -554,6 +587,218 @@ def new(self):
...
+            cmd += [filepath]
+            self.kernel.shell.run_line_magic('load', ' '.join(cmd))
+            self.ipython_cell_executed()
+            self.update_recent_script_list(filepath)
+        except Exception as e:
+            QMessageBox.critical(self, "Error", "Cannot load script file {}:\n{}"
+                                 .format(os.path.basename(filepath), e))
+
+    def load_script(self, filepath=None):
+        if self._ask_to_save_if_unsaved_modifications():
```
Comment (alixdamman) [inline larray_editor/editor.py]: Still need to be changed

File: larray_editor/editor.py [hunk 3]
```diff
@@ -554,6 +586,177 @@ def new(self):
...
+                    pass
+            except Exception as e:
+                QMessageBox.critical(self, "Error", "Cannot load script file {}:\n{}"
+                                     .format(os.path.basename(filepath), e))
+        else:
+            QMessageBox.critical(self, "Error", "Cannot load script file {}. You need to install ipython and "\
+                                                "qtconsole.".format(os.path.basename(filepath)))
+
+    def load_script(self, filepath=None):
+        if self._ask_to_save_if_unsaved_modifications():
```
[the remark this answers is not in the document]
Comment (alixdamman) [reply larray_editor/editor.py]: I suggest to add a CheckBox "Reset session before to load" and clearly explain what "Load Script" does (inject code in the current session).

[the remark this answers is not in the document]
Comment (gdementen) [reply larray_editor/editor.py]: Yes, that seems like a good idea. Isn't "clear" more intuitive than reset?

File: larray_editor/editor.py (hunk 2 above)
Comment (alixdamman) [reply larray_editor/editor.py]: remainder: add a CheckBox "Clear session before to load"

**Comment by alixdamman** [comment]
@gdementen We add it to version 0.27 or we wait for 0.28?

**Comment by gdementen** [comment]
How confident are you about this?

**Comment by alixdamman** [comment]
@gdementen Only "Save Executed Commands To Script" is available in the File menu. I also use 2 radio buttons to choose between "overwrite file" and "append to file".

File: larray_editor/editor.py [hunk 4]
```diff
@@ -567,6 +614,236 @@ def new(self):
...
+        group_box_layout = QGridLayout()
+        # all lines
+        radio_button_all_lines = QRadioButton("Save all history")
+        radio_button_all_lines.setChecked(True)
+        group_box_layout.addWidget(radio_button_all_lines, 0, 0)
+        # specific lines
+        radio_button_specific_lines = QRadioButton("Save input lines")
+        radio_button_specific_lines.setToolTip("Selected (ranges of) input lines must be separated with whitespaces.\n"
+                                               "Ranges could be specified as x-y (x..y) or in python-style "
+                                               "x:y (x..(y-1)).")
```
Comment (gdementen) [inline larray_editor/editor.py]: Change:
> "Ranges could be specified as x-y (x..y) or in python-style "
> "x:y (x..(y-1))."

to:

> "Ranges should be specified as x..y (or x-y) or in python-style "
> "x:y (equivalent to x..(y-1))."

?

**Comment by gdementen** [comment]
@alixdamman we need to merge this before the release (it already appears in the changelog)

**Comment by alixdamman** [comment]
@gdementen you can check the last commit

**Pull request merged**